plugin/trino-kafka/src/main/java/io/trino/plugin/kafka/schema/confluent/AbstractConfluentRowEncoder.java

```diff
@@ -0,0 +1,246 @@
+        this.topic = requireNonNull(topic, "topic is null");
+    }
+
+    protected abstract Object buildRow();
+
+    protected byte[] serialize(Object object)
+    {
+        return kafkaAvroSerializer.serialize(topic, object);
+    }
+
+    protected abstract void addValue(Block block, int position);
+
+    @Override
+    public void appendColumnValue(Block block, int position)
+    {
+        checkArgument(getCurrentColumnIndex() < columnHandles.size(), format("currentColumnIndex '%d' is greater than number of columns '%d'", getCurrentColumnIndex(), columnHandles.size()));
+        addValue(block, position);
+        currentColumnIndex++;
+    }
+
+    protected Object getValue(Type type, Block block, int position, Schema fieldSchema)
+    {
+        if (block.isNull(position)) {
+            checkState(fieldSchema.isNullable(), "Unexpected null value for non-nullable schema '%s'", fieldSchema.toString(true));
+            return null;
+        }
+        // Since the value is non-null, if this a union, extract complex type
+        fieldSchema = extractBaseType(fieldSchema);
+
+        if (type == BOOLEAN) {
+            return type.getObjectValue(session, block, position);
+        }
+        else if (type == BIGINT) {
+            return type.getLong(block, position);
+        }
+        else if (type == INTEGER) {
+            return toIntExact(type.getLong(block, position));
+        }
+        else if (type == SMALLINT) {
+            return Shorts.checkedCast(type.getLong(block, position));
+        }
+        else if (type == TINYINT) {
+            return SignedBytes.checkedCast(type.getLong(block, position));
+        }
+        else if (type == DOUBLE) {
+            return type.getDouble(block, position);
+        }
+        else if (type == REAL) {
+            return intBitsToFloat(toIntExact(type.getLong(block, position)));
+        }
+        else if (type instanceof VarcharType) {
+            if (fieldSchema.getType() == ENUM) {
+                return new GenericData.EnumSymbol(fieldSchema, type.getSlice(block, position).toStringUtf8());
+            }
+            return type.getSlice(block, position).toStringUtf8();
+        }
+        else if (type instanceof VarbinaryType) {
+            return type.getSlice(block, position).toByteBuffer();
+        }
+
+        if (type instanceof ArrayType) {
+            ArrayType arrayType = (ArrayType) type;
+            Type elementType = arrayType.getElementType();
+            Block arrayBlock = block.getObject(position, Block.class);
+            List<Object> list = new ArrayList<>(arrayBlock.getPositionCount());
+            Schema elementAvroType = fieldSchema.getElementType();
+            for (int i = 0; i < arrayBlock.getPositionCount(); i++) {
+                list.add(getValue(elementType, arrayBlock, i, elementAvroType));
+            }
+            return Collections.unmodifiableList(list);
```

Comment:
specific reason to avoid Guava collections in this class?

Reply:
Switching to guava:) Saw the suggestion to use guava since if we use it everywhere then we avoid extra copies and also preserve order for builders, etc.

Reply:
Update: had to switch back because lists and maps may contain null values, for maps it may be a non-null key with a null value.